File: server/chat_app_server/app/api/messages.py
# 消息相关API路由

from fastapi import APIRouter, HTTPException, Query
from typing import Optional, List, Dict, Any
import logging
import json

from app.models.message import MessageCreate, row_to_dict

logger = logging.getLogger(__name__)
router = APIRouter()


def format_message(message_row) -> Dict[str, Any]:
    """格式化消息数据"""
    if not message_row:
        return None
    
    message_dict = row_to_dict(message_row)
    # 解析JSON字段
    if message_dict.get('tool_calls'):
        message_dict['tool_calls'] = json.loads(message_dict['tool_calls'])
    if message_dict.get('metadata'):
        message_dict['metadata'] = json.loads(message_dict['metadata'])
    
    return message_dict


@router.get("/sessions/{session_id}/messages")
async def get_session_messages(session_id: str, limit: Optional[int] = Query(None, ge=1, le=1000)):
    """获取会话的消息列表"""
    try:
        messages = await MessageCreate.get_by_session(session_id, limit=limit)
        logger.info(f"获取会话 {session_id} 的 {len(messages)} 条消息")
        return messages
        
    except Exception as e:
        logger.error(f"获取会话消息失败: {e}")
        raise HTTPException(status_code=500, detail="获取会话消息失败")

@router.post("/sessions/{session_id}/messages")
async def create_message(session_id: str, message: MessageCreate):
    """创建新消息"""
    try:
        # 确保session_id一致
        message.session_id = session_id
        
        new_message = await MessageCreate.create(message)
        logger.info(f"创建消息成功: {new_message['id']}")
        return new_message
        
    except Exception as e:
        logger.error(f"创建消息失败: {e}")
        raise HTTPException(status_code=500, detail="创建消息失败")

File: server/chat_app_node_server/src/api/messages.js
/**
 * Messages API 路由 - 消息管理
 * 复刻自 Python: app/api/messages.py
 */

import express from 'express';
import { MessageService } from '../models/message.js';
import { logger } from '../utils/logger.js';

const router = express.Router();

/**
 * GET /api/messages - 获取消息列表
 * 支持按 session_id 过滤
 */
router.get('/', async (req, res) => {
  try {
    const { session_id, limit, offset } = req.query;

    if (!session_id) {
      return res.status(400).json({ error: '必须提供 session_id' });
    }

    const messages = await MessageService.getBySession(
      session_id,
      limit ? parseInt(limit, 10) : null,
      offset ? parseInt(offset, 10) : 0
    );

    res.json(messages);
  } catch (error) {
    logger.error('获取消息列表失败:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * POST /api/messages - 创建新消息
 */
router.post('/', async (req, res) => {
  try {
    const { sessionId, role, content, toolCalls, tool_call_id, reasoning, metadata } = req.body;

    if (!sessionId || !role || !content) {
      return res.status(400).json({ error: 'sessionId, role 和 content 不能为空' });
    }

    const message = await MessageService.create({
      sessionId,
      role,
      content,
      toolCalls,
      tool_call_id,
      reasoning,
      metadata
    });

    res.status(201).json(message);
  } catch (error) {
    logger.error('创建消息失败:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * GET /api/messages/:id - 获取特定消息
 */
router.get('/:id', async (req, res) => {
  try {
    const message = await MessageService.getById(req.params.id);

    if (!message) {
      return res.status(404).json({ error: '消息不存在' });
    }

    res.json(message);
  } catch (error) {
    logger.error('获取消息失败:', error);
    res.status(500).json({ error: error.message });
  }
});

/**
 * DELETE /api/messages/:id - 删除消息
 */
router.delete('/:id', async (req, res) => {
  try {
    await MessageService.delete(req.params.id);

    res.json({ success: true, message: '消息已删除' });
  } catch (error) {
    logger.error('删除消息失败:', error);
    res.status(500).json({ error: error.message });
  }
});

export default router;
